File: src/main/java/pers/neige/neigeitems/action/node/impl/WeightParser.java
package pers.neige.neigeitems.action.node.impl;

import lombok.NonNull;
import lombok.val;
import org.jetbrains.annotations.Nullable;
import pers.neige.neigeitems.action.ActionContext;
import pers.neige.neigeitems.action.node.NodeParser;
import pers.neige.neigeitems.config.ConfigReader;
import pers.neige.neigeitems.manager.BaseActionManager;
import pers.neige.neigeitems.utils.NumberParser;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.Collectors;

public class WeightParser extends NodeParser {
    public WeightParser(@NonNull BaseActionManager manager) {
        super(manager);
    }

    @Override
    public @NonNull String getId() {
        return "weight";
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull ConfigReader params
    ) {
        return handle(
            params.getStringList("values").stream().map(it -> this.manager.parseNode(it, context)).collect(Collectors.toList())
        );
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull List<String> params
    ) {
        return handle(
            params
        );
    }

    private @Nullable String handle(
        @Nullable List<String> values
    ) {
        if (values == null || values.isEmpty()) return null;
        val info = new HashMap<String, BigDecimal>();
        BigDecimal total = BigDecimal.ZERO;
        // 加载所有参数并遍历
        for (val value : values) {
            // 检测权重
            val index = value.indexOf("::");
            // 无权重, 直接记录
            if (index == -1) {
                info.put(value, info.getOrDefault(value, BigDecimal.ZERO).add(BigDecimal.ONE));
                total = total.add(BigDecimal.ONE);
            } else {
                // 有权重, 根据权重大小进行记录
                val weight = NumberParser.parseBigDecimal(value.substring(0, index), BigDecimal.ONE);
                val string = value.substring(index + 2);
                info.put(string, info.getOrDefault(string, BigDecimal.ZERO).add(weight));
                total = total.add(weight);
            }
        }
        val random = BigDecimal.valueOf(ThreadLocalRandom.current().nextDouble()).multiply(total);
        BigDecimal current = BigDecimal.ZERO;
        for (val entry : info.entrySet()) {
            current = current.add(entry.getValue());
            if (random.compareTo(current) <= 0) {
                return entry.getKey();
            }
        }
        return null;
    }
}

File: src/main/kotlin/pers/neige/neigeitems/section/impl/WeightParser.kt
package pers.neige.neigeitems.section.impl

import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import pers.neige.neigeitems.section.SectionParser
import pers.neige.neigeitems.utils.SectionUtils.parseSection
import java.math.BigDecimal
import java.util.concurrent.ThreadLocalRandom

/**
 * 权重节点解析器
 */
object WeightParser : SectionParser() {
    override val id: String = "weight"

    override fun onRequest(
        data: ConfigurationSection,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?, sections:
        ConfigurationSection?
    ): String? {
        return handler(
            cache,
            player,
            sections,
            true,
            data.getStringList("values")
        )
    }

    override fun onRequest(
        args: List<String>,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String {
        return handler(
            cache,
            player,
            sections,
            false,
            args
        ) ?: "<$id::${args.joinToString("_")}>"
    }

    /**
     * @param cache 解析值缓存
     * @param player 待解析玩家
     * @param sections 节点池
     * @param parse 是否对参数进行节点解析
     * @param values 文本列表
     * @return 解析值
     */
    private fun handler(
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?,
        parse: Boolean,
        values: List<String>
    ): String? {
        val info = HashMap<String, BigDecimal>()
        var total = BigDecimal.ZERO
        // 加载所有参数并遍历
        values.forEach {
            val value = it.parseSection(parse, cache, player, sections)
            // 检测权重
            when (val index = value.indexOf("::")) {
                // 无权重, 直接记录
                -1 -> {
                    info[value]?.let {
                        info[value] = it.add(BigDecimal.ONE)
                    } ?: let { info[value] = BigDecimal.ONE }
                    total = total.add(BigDecimal.ONE)
                }
                // 有权重, 根据权重大小进行记录
                else -> {
                    val weight = value.substring(0, index).toBigDecimalOrNull() ?: BigDecimal.ONE
                    val string = value.substring(index+2, value.length)
                    info[string]?.let {
                        info[string] = it.add(weight)
                    } ?: let { info[string] = weight }
                    total = total.add(weight)
                }
            }
        }
        // 根据最后的记录值进行字符随机
        return when {
            info.isEmpty() -> null
            else -> {
                val random = BigDecimal(ThreadLocalRandom.current().nextDouble().toString()).multiply(total)
                var current = BigDecimal.ZERO
                var result: String? = null
                for ((key, value) in info) {
                    current = current.add(value)
                    if (random <= current) {
                        result = key.parseSection(parse, cache, player, sections)
                        break
                    }
                }
                result
            }
        }
    }
}